### ISSUE TITLE
ConcurrentModificationException when setting properties while creating indexes

### ISSUE THREAD

**Issue by cmorgner** [neo4j/neo4j#12003]
A set property call via the Bolt Java driver sometimes fails, causing the transaction to be rolled back when indexes are created in a separate thread. The debug.log then shows a `ConcurrentModificationException`. See stack trace below.

**Neo4j Version:** 3.4.4
**Operating System:** Kubuntu 18.04
**API:** Bolt Java driver

### Steps to reproduce
I was not able to create an isolated test that provokes the error.

### Expected behavior
No error, transaction is committed.

### Actual behavior
Transaction fails, Bolt driver throws the following exception:
```
Aug 02, 2018 5:13:10 PM org.neo4j.driver.internal.logging.JULogger error
SEVERE: [0x1fae42fc] Fatal error occurred in the pipeline
org.neo4j.driver.v1.exceptions.DatabaseException: null
        at org.neo4j.driver.internal.util.ErrorUtil.newNeo4jError(ErrorUtil.java:66)
        at org.neo4j.driver.internal.async.inbound.InboundMessageDispatcher.handleFailureMessage(InboundMessageDispatcher.java:137)
        at org.neo4j.driver.internal.messaging.PackStreamMessageFormatV1$ReaderV1.unpackFailureMessage(PackStreamMessageFormatV1.java:336)
        at org.neo4j.driver.internal.messaging.PackStreamMessageFormatV1$ReaderV1.read(PackStreamMessageFormatV1.java:300)
        at org.neo4j.driver.internal.async.inbound.InboundMessageHandler.channelRead0(InboundMessageHandler.java:82)
        at org.neo4j.driver.internal.async.inbound.InboundMessageHandler.channelRead0(InboundMessageHandler.java:34)
        at org.neo4j.driver.internal.shaded.io.netty.channel.SimpleChannelInboundHandler.channelRead(SimpleChannelInboundHandler.java:105)
        at org.neo4j.driver.internal.shaded.io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
        at org.neo4j.driver.internal.shaded.io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
        at org.neo4j.driver.internal.shaded.io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340)
        at org.neo4j.driver.internal.shaded.io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:310)
        at org.neo4j.driver.internal.shaded.io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:284)
        at org.neo4j.driver.internal.async.inbound.MessageDecoder.channelRead(MessageDecoder.java:39)
        at org.neo4j.driver.internal.shaded.io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
        at org.neo4j.driver.internal.shaded.io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
        at org.neo4j.driver.internal.shaded.io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340)
        at org.neo4j.driver.internal.shaded.io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:310)
        at org.neo4j.driver.internal.shaded.io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:297)
        at org.neo4j.driver.internal.shaded.io.netty.handler.codec.ByteToMessageDecoder.callDecode(ByteToMessageDecoder.java:413)
        at org.neo4j.driver.internal.shaded.io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:265)
        at org.neo4j.driver.internal.shaded.io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
        at org.neo4j.driver.internal.shaded.io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
        at org.neo4j.driver.internal.shaded.io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340)
        at org.neo4j.driver.internal.shaded.io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1414)
        at org.neo4j.driver.internal.shaded.io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
        at org.neo4j.driver.internal.shaded.io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
        at org.neo4j.driver.internal.shaded.io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:945)
        at org.neo4j.driver.internal.shaded.io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:146)
        at org.neo4j.driver.internal.shaded.io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:645)
        at org.neo4j.driver.internal.shaded.io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:580)
        at org.neo4j.driver.internal.shaded.io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:497)
        at org.neo4j.driver.internal.shaded.io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:459)
        at org.neo4j.driver.internal.shaded.io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:886)
        at org.neo4j.driver.internal.shaded.io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
        at java.lang.Thread.run(Thread.java:748)
```

#### Contents of logs/debug.log
```
[...]
2018-08-02 15:13:10.094+0000 INFO [o.n.k.i.a.i.IndexPopulationJob] Completed node store scan. Flushing all pending updates.
BatchingMultipleIndexPopulator{activeTasks=1, executor=java.util.concurrent.ThreadPoolExecutor@11188729[Running, pool size = 11, active threads = 1, queued tasks = 0, completed tasks = 18], batchedUpdates = [0 updates, 0 updates, 0 updates, 0 updates, 0 updates, 0 updates, 0 updates, 0 updates, 0 updates, 0 updates, 0 updates, 0 updates, 0 updates, 0 updates, 0 updates, 0 updates, 0 updates, 0 updates, 0 updates], queuedUpdates = 0}
2018-08-02 15:13:10.094+0000 INFO [o.n.k.i.a.i.IndexPopulationJob] Shutting down executor.
BatchingMultipleIndexPopulator{activeTasks=6, executor=java.util.concurrent.ThreadPoolExecutor@11188729[Running, pool size = 11, active threads = 0, queued tasks = 0, completed tasks = 33], batchedUpdates = [0 updates, 0 updates, 0 updates, 0 updates, 0 updates, 0 updates, 0 updates, 0 updates, 0 updates, 0 updates, 0 updates, 0 updates, 0 updates, 0 updates, 0 updates, 0 updates, 0 updates, 0 updates, 0 updates], queuedUpdates = 0}
2018-08-02 15:13:10.100+0000 ERROR [o.n.b.v.r.ErrorReporter] Client triggered an unexpected error [Neo.DatabaseError.General.UnknownError]: null, reference 85bae31b-3ea9-42d0-8cfe-78a6e7cde651.
2018-08-02 15:13:10.101+0000 ERROR [o.n.b.v.r.ErrorReporter] Client triggered an unexpected error [Neo.DatabaseError.General.UnknownError]: null, reference 85bae31b-3ea9-42d0-8cfe-78a6e7cde651.
java.util.ConcurrentModificationException
	at java.util.ArrayList$Itr.checkForComodification(ArrayList.java:907)
	at java.util.ArrayList$Itr.next(ArrayList.java:857)
	at org.neo4j.kernel.impl.newapi.NodeSchemaMatcher.onMatchingSchema(NodeSchemaMatcher.java:72)
	at org.neo4j.kernel.impl.newapi.IndexTxStateUpdater.onPropertyChange(IndexTxStateUpdater.java:149)
	at org.neo4j.kernel.impl.newapi.Operations.nodeSetProperty(Operations.java:527)
	at org.neo4j.cypher.internal.runtime.interpreted.TransactionBoundQueryContext$NodeOperations.setProperty(TransactionBoundQueryContext.scala:443)
	at org.neo4j.cypher.internal.compatibility.v3_4.ExceptionTranslatingQueryContext$ExceptionTranslatingOperations$$anonfun$setProperty$1.apply$mcV$sp(ExceptionTranslatingQueryContext.scala:284)
	at org.neo4j.cypher.internal.compatibility.v3_4.ExceptionTranslatingQueryContext$ExceptionTranslatingOperations$$anonfun$setProperty$1.apply(ExceptionTranslatingQueryContext.scala:284)
	at org.neo4j.cypher.internal.compatibility.v3_4.ExceptionTranslatingQueryContext$ExceptionTranslatingOperations$$anonfun$setProperty$1.apply(ExceptionTranslatingQueryContext.scala:284)
	at org.neo4j.cypher.internal.compatibility.v3_4.ExceptionTranslationSupport$class.translateException(ExceptionTranslationSupport.scala:33)
	at org.neo4j.cypher.internal.compatibility.v3_4.ExceptionTranslatingQueryContext.translateException(ExceptionTranslatingQueryContext.scala:41)
	at org.neo4j.cypher.internal.compatibility.v3_4.ExceptionTranslatingQueryContext$ExceptionTranslatingOperations.setProperty(ExceptionTranslatingQueryContext.scala:284)
	at org.neo4j.cypher.internal.runtime.interpreted.UpdateCountingQueryContext$CountingOps.setProperty(UpdateCountingQueryContext.scala:191)
	at org.neo4j.cypher.internal.runtime.interpreted.pipes.SetPropertyFromMapOperation$$anonfun$setPropertiesFromMap$3.apply(SetOperation.scala:192)
	at org.neo4j.cypher.internal.runtime.interpreted.pipes.SetPropertyFromMapOperation$$anonfun$setPropertiesFromMap$3.apply(SetOperation.scala:188)
	at scala.collection.TraversableLike$WithFilter$$anonfun$foreach$1.apply(TraversableLike.scala:733)
	at scala.collection.immutable.Map$Map3.foreach(Map.scala:161)
	at scala.collection.TraversableLike$WithFilter.foreach(TraversableLike.scala:732)
	at org.neo4j.cypher.internal.runtime.interpreted.pipes.SetPropertyFromMapOperation.setPropertiesFromMap(SetOperation.scala:188)
	at org.neo4j.cypher.internal.runtime.interpreted.pipes.SetPropertyFromMapOperation.set(SetOperation.scala:175)
	at org.neo4j.cypher.internal.runtime.interpreted.pipes.SetPipe$$anonfun$internalCreateResults$1.apply(SetPipe.scala:30)
	at org.neo4j.cypher.internal.runtime.interpreted.pipes.SetPipe$$anonfun$internalCreateResults$1.apply(SetPipe.scala:29)
	at scala.collection.Iterator$$anon$11.next(Iterator.scala:410)
	at org.neo4j.cypher.internal.runtime.interpreted.pipes.EmptyResultPipe.internalCreateResults(EmptyResultPipe.scala:29)
	at org.neo4j.cypher.internal.runtime.interpreted.pipes.PipeWithSource.createResults(Pipe.scala:76)
	at org.neo4j.cypher.internal.runtime.interpreted.pipes.PipeWithSource.createResults(Pipe.scala:72)
	at org.neo4j.cypher.internal.compatibility.v3_4.runtime.executionplan.BaseExecutionResultBuilderFactory$BaseExecutionWorkflowBuilder.createResults(DefaultExecutionResultBuilderFactory.scala:105)
	at org.neo4j.cypher.internal.compatibility.v3_4.runtime.executionplan.BaseExecutionResultBuilderFactory$BaseExecutionWorkflowBuilder.build(DefaultExecutionResultBuilderFactory.scala:77)
	at org.neo4j.cypher.internal.compatibility.v3_4.runtime.BuildInterpretedExecutionPlan$$anonfun$getExecutionPlanFunction$1.apply(BuildInterpretedExecutionPlan.scala:97)
	at org.neo4j.cypher.internal.compatibility.v3_4.runtime.BuildInterpretedExecutionPlan$$anonfun$getExecutionPlanFunction$1.apply(BuildInterpretedExecutionPlan.scala:80)
	at org.neo4j.cypher.internal.compatibility.v3_4.runtime.BuildInterpretedExecutionPlan$InterpretedExecutionPlan.run(BuildInterpretedExecutionPlan.scala:111)
	at org.neo4j.cypher.internal.compatibility.LatestRuntimeVariablePlannerCompatibility$ExecutionPlanWrapper$$anonfun$run$1.apply(LatestRuntimeVariablePlannerCompatibility.scala:128)
	at org.neo4j.cypher.internal.compatibility.LatestRuntimeVariablePlannerCompatibility$ExecutionPlanWrapper$$anonfun$run$1.apply(LatestRuntimeVariablePlannerCompatibility.scala:124)
	at org.neo4j.cypher.exceptionHandler$runSafely$.apply(exceptionHandler.scala:89)
	at org.neo4j.cypher.internal.compatibility.LatestRuntimeVariablePlannerCompatibility$ExecutionPlanWrapper.run(LatestRuntimeVariablePlannerCompatibility.scala:124)
	at org.neo4j.cypher.internal.PreparedPlanExecution.execute(PreparedPlanExecution.scala:29)
	at org.neo4j.cypher.internal.ExecutionEngine.execute(ExecutionEngine.scala:119)
	at org.neo4j.cypher.internal.javacompat.ExecutionEngine.executeQuery(ExecutionEngine.java:61)
	at org.neo4j.bolt.v1.runtime.TransactionStateMachineSPI$1.start(TransactionStateMachineSPI.java:144)
	at org.neo4j.bolt.v1.runtime.TransactionStateMachine$State.startExecution(TransactionStateMachine.java:444)
	at org.neo4j.bolt.v1.runtime.TransactionStateMachine$State$2.run(TransactionStateMachine.java:341)
	at org.neo4j.bolt.v1.runtime.TransactionStateMachine.run(TransactionStateMachine.java:81)
	at org.neo4j.bolt.v1.runtime.BoltStateMachine$State$2.run(BoltStateMachine.java:457)
	at org.neo4j.bolt.v1.runtime.BoltStateMachine.run(BoltStateMachine.java:225)
	at org.neo4j.bolt.v1.messaging.BoltMessageRouter.lambda$onRun$3(BoltMessageRouter.java:93)
	at org.neo4j.bolt.runtime.DefaultBoltConnection.processNextBatch(DefaultBoltConnection.java:195)
	at org.neo4j.bolt.runtime.DefaultBoltConnection.processNextBatch(DefaultBoltConnection.java:143)
	at org.neo4j.bolt.runtime.ExecutorBoltScheduler.executeBatch(ExecutorBoltScheduler.java:170)
	at org.neo4j.bolt.runtime.ExecutorBoltScheduler.lambda$scheduleBatchOrHandleError$2(ExecutorBoltScheduler.java:153)
	at java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1590)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
2018-08-02 15:13:10.112+0000 INFO [o.n.k.i.a.i.IndexPopulationJob] Index population started: [:DOMNode(showForLocales) [provider: {key=lucene+native, version=2.0}]]
2018-08-02 15:13:10.112+0000 INFO [o.n.k.i.a.i.IndexPopulationJob] Index population started: [:SourcePattern(type) [provider: {key=lucene+native, version=2.0}]]
2018-08-02 15:13:10.115+0000 INFO [o.n.k.i.a.i.IndexPopulationJob] Index creation finished. Index [:Language(hidden) [provider: {key=lucene+native, version=2.0}]] is ONLINE.
2018-08-02 15:13:10.115+0000 INFO [o.n.k.i.a.i.IndexPopulationJob] Index creation finished. Index [:ShadowDocument(lastModifiedDate) [provider: {key=lucene+native, version=2.0}]] is ONLINE.
2018-08-02 15:13:10.116+0000 INFO [o.n.k.i.a.i.IndexPopulationJob] Index creation finished. Index [:Image(checksum) [provider: {key=lucene+native, version=2.0}]] is ONLINE.
2018-08-02 15:13:10.120+0000 INFO [o.n.k.i.a.i.IndexPopulationJob] Index creation finished. Index [:DataSource(name) [provider: {key=lucene+native, version=2.0}]] is ONLINE.
2018-08-02 15:13:10.124+0000 INFO [o.n.k.i.a.i.IndexPopulationJob] Index population started: [:GraphObject(visibleToPublicUsers) [provider: {key=lucene+native, version=2.0}]]
2018-08-02 15:13:10.145+0000 INFO [o.n.k.i.a.i.IndexPopulationJob] Index creation finished. Index [:XMPPRequest(visibleToAuthenticatedUsers) [provider: {key=lucene+native, version=2.0}]] is ONLINE.
```

**Comment by chrisvest** [comment]
Yeah, that's definitely a bug. Thanks for reporting.

**Comment by chrisvest** [comment]
A fix for this will be included in the next 3.4 release. Closing.
